src/common/guac_rect.c

```diff
@@ -80,3 +80,187 @@ void guac_common_rect_constrain(guac_common_rect* rect, const guac_common_rect*
 
 }
 
+int guac_common_rect_adjust(int divisor, guac_common_rect* rect,
+                            const guac_common_rect* max) {
+
+    /* Invalid divisor received */
+    if (divisor <= 0)
+        return -1;
+
+    /* Calculate how much the rectangle must be adjusted to not produce a
+     * remainder if divided by the divisor. */
+    int dw = divisor - rect->width % divisor;
+    int dh = divisor - rect->height % divisor;
+
+    int dx = dw / 2;
+    int dy = dh / 2;
+
+    /* Set initial extents of adjusted rectangle. */
+    int top = rect->y - dy;
+    int left = rect->x - dx;
+    int bottom = top + rect->height + dh;
+    int right = left + rect->width + dw;
+
+    /* The max rectangle (clipping rectangle) */
+    int max_left   = max->x;
+    int max_top    = max->y;
+    int max_right  = max_left + max->width;
+    int max_bottom = max_top  + max->height;
+
+    /* If the adjusted rectangle has sides beyond the max rectangle, or is larger
+     * in any direction; shift or adjust the rectangle while trying to
+     * maintain it's quotient without a remainder.
+     */
+
+    /* Adjust left/right */
+    if (right > max_right) {
+
+        /* shift to left */
+        dw = right - max_right;
+        right -= dw;
+        left -= dw;
+
+        /* clamp left if too far */
+        if (left < max_left) {
+            left = max_left;
+        }
+    }
+    else if (left < max_left) {
+
+        /* shift to right */
+        dw = max_left - left;
+        left += dw;
+        right += dw;
+
+        /* clamp right if too far */
+        if (right > max_right) {
+            right = max_right;
+        }
+    }
+
+    /* Adjust top/bottom */
+    if (bottom > max_bottom) {
+
+        /* shift up */
+        dh = bottom - max_bottom;
+        bottom -= dh;
+        top -= dh;
+
+        /* clamp top if too far */
+        if (top < max_top) {
+            top = max_top;
+        }
+    }
+    else if (top < max_top) {
+
+        /* shift down */
+        dh = max_top - top;
+        top += dh;
+        bottom += dh;
+
+        /* clamp bottom if too far */
+        if (bottom > max_bottom) {
+            bottom = max_bottom;
+        }
+    }
+
+    /* Commit rect */
+    guac_common_rect_init(rect, left, top, right - left, bottom - top);
+
+    return 0;
+
+}
+
+int guac_common_rect_intersects(guac_common_rect* rect, const guac_common_rect* min) {
+
+    /* Empty (no intersection) */
+    if (min->x + min->width < rect->x || rect->x + rect->width < min->x ||
+        min->y + min->height < rect->y || rect->y + rect->height < min->y) {
+        return -1;
+    }
+    /* Complete */
+    else if (min->x <= rect->x && (min->x + min->width) >= (rect->x + rect->width) &&
+        min->y <= rect->y && (min->y + min->height) >= (rect->y + rect->height)) {
+        return 1;
+    }
+    /* Partial intersection */
+    else {
+        return 0;
```

Comment:
Rather than having one non-zero value being "no intersection", another non-zero value being "complete intersection", and zero being "partial intersection", we should probably use zero for the "no intersection" case. That way, *all* types of intersection are non-zero values (which will make `if` happy), and those that want to know which kind of intersection can may further attention to the value.